spark/src/test/scala/org/apache/spark/sql/delta/IdentityColumnDMLSuite.scala

```diff
@@ -0,0 +1,591 @@
+          .coalesce(1)
+          .write
+          .format("delta")
+          .mode("overwrite")
+          .option(DeltaOptions.REPLACE_WHERE_OPTION, "is_even = true")
+          .saveAsTable(table)
+
+        highWaterMarkFromData = validateIdentity(
+          table,
+          expectedRowCount = 4,
+          start = start,
+          step = step,
+          minValue = 10,
+          maxValue = 12,
+          oldHighWaterMark = highWaterMarkFromData)
+        assert(highWaterMarkFromData === highWatermarkFromDeltaLog())
+        // IDENTITY values for not-updated shouldn't change.
+        checkAnswer(
+          sql(s"SELECT id FROM $table WHERE is_odd = true"),
+          idBeforeReplaceWhere2
+        )
+
+        // IDENTITY VALUES for inserted change records and data should be consistent.
+        checkAnswer(
+          sql(s"SELECT id FROM table_changes('$table', 3, 3) " +
+            "WHERE is_even = true and _change_type = 'insert'"),
+          sql(s"SELECT id FROM $table WHERE is_even = true")
+        )
+
+        // ReplaceWhere source data contains an Identity Column will be blocked.
+        val e = intercept[AnalysisException] {
+          Seq((15, 14), (17, 16)).toDF("id", "value")
+            .withColumn("is_odd", $"value" % 2 =!= 0)
+            .withColumn("is_even", $"value" % 2 === 0)
+            .coalesce(1)
+            .write
+            .format("delta")
+            .mode("overwrite")
+            .option(DeltaOptions.REPLACE_WHERE_OPTION, "is_even = true")
+            .saveAsTable(table)
+        }
+        assert(e.getMessage.contains(
+          "Providing values for GENERATED ALWAYS AS IDENTITY column id is not supported."))
+      }
+    }
+  }
+
+  Seq(true, false).foreach { isPartitioned =>
+    test(s"replaceWhere with CDF enabled [isPartitioned: $isPartitioned]") {
+      testReplaceWhereWithCDF(isPartitioned)
+    }
+  }
+
+  test("CDF for tables with identity column - MERGE") {
+    val src = "identity_test_src"
+    val tgt = "identity_test_tgt"
+    withTable(tgt) {
+      generateTableWithIdentityColumn(tgt)
+      sql(s"ALTER TABLE $tgt SET TBLPROPERTIES (${DeltaConfigs.CHANGE_DATA_FEED.key}=true)")
+
+      withTempView(src) {
+        (1 :: 20 :: 30 :: Nil).toDF("value").createOrReplaceTempView(src)
+        sql(
+          s"""MERGE INTO $tgt target
+             |USING $src src ON target.value = src.value
+             |WHEN MATCHED THEN UPDATE SET target.value = src.value
+             |WHEN NOT MATCHED THEN INSERT (target.value) VALUES (src.value)""".stripMargin)
+      }
+
+      val result = sql(
+        s"""SELECT id, value
+           |FROM $tgt
+           |ORDER BY id""".stripMargin)
+        .as[IdentityColumnTestTableRow]
+        .collect()
+
+      assert(result.length === 8)
+      checkGeneratedIdentityValues(
+        sortedRows = result,
+        start = 0,
+        step = 1,
+        expectedLowerBound = 0,
+        expectedUpperBound =
```

Comment:
Let's use a hard coded value for expectedUpperBound. When we dynamically compute expected test result, we risk having a testing error where the test is erroneously passing due to the dynamic computation also being wrong.  

We just want it to be in a "reasonable" range. E.g. 0 - 20 or 0 - 100.